src/realm/sync/noinst/client_reset.cpp

```diff
@@ -499,8 +509,995 @@ void EmbeddedObjectConverter::process_pending()
+{
+    // For link columns, `Obj::get_listbase_ptr()` always returns an instance whose concrete type is
+    // `LnkLst`, which uses condensed indexes. However, we are interested in using non-condensed
+    // indexes, so we need to manually construct a `Lst<ObjKey>` instead for lists of non-embedded
+    // links.
+    REALM_ASSERT(col.is_list());
+    std::unique_ptr<LstBase> list;
+    if (col.get_type() == col_type_Link || col.get_type() == col_type_LinkList) {
+        auto table = obj.get_table();
+        if (!table->get_link_target(col)->is_embedded()) {
+            list = obj.get_list_ptr<ObjKey>(col);
+        }
+        else {
+            list = obj.get_listbase_ptr(col);
+        }
+    }
+    else {
+        list = obj.get_listbase_ptr(col);
+    }
+    return list;
+}
+
+struct InternDictKey {
+    size_t pos = realm::npos;
+    size_t size = realm::npos;
+    bool is_null() const
+    {
+        return pos == realm::npos && size == realm::npos;
+    }
+    constexpr bool operator==(const InternDictKey& other) const noexcept
+    {
+        return pos == other.pos && size == other.size;
+    }
+    constexpr bool operator!=(const InternDictKey& other) const noexcept
+    {
+        return !operator==(other);
+    }
+    constexpr bool operator<(const InternDictKey& other) const noexcept
+    {
+        if (pos < other.pos) {
+            return true;
+        }
+        else if (pos == other.pos) {
+            return size < other.size;
+        }
+        return false;
+    }
+};
+
+struct InterningBuffer {
+    StringData get_key(const InternDictKey& key) const
+    {
+        if (key.is_null()) {
+            return {};
+        }
+        if (key.size == 0) {
+            return "";
+        }
+        REALM_ASSERT(key.pos < m_dict_keys_buffer.size());
+        REALM_ASSERT(key.pos + key.size <= m_dict_keys_buffer.size());
+        return StringData{m_dict_keys_buffer.data() + key.pos, key.size};
+    }
+
+    InternDictKey get_or_add(const StringData& str)
+    {
+        for (auto& key : m_dict_keys) {
+            StringData existing = get_key(key);
+            if (existing == str) {
+                return key;
+            }
+        }
+        InternDictKey new_key{};
+        if (str.is_null()) {
+            m_dict_keys.push_back(new_key);
+        }
+        else {
+            size_t next_pos = m_dict_keys_buffer.size();
+            new_key.pos = next_pos;
+            new_key.size = str.size();
+            m_dict_keys_buffer.append(str);
+            m_dict_keys.push_back(new_key);
+        }
+        return new_key;
+    }
+    InternDictKey get_interned_key(const StringData& str) const
+    {
+        if (str.is_null()) {
+            return {};
+        }
+        for (auto& key : m_dict_keys) {
+            StringData existing = get_key(key);
+            if (existing == str) {
+                return key;
+            }
+        }
+        REALM_UNREACHABLE();
```

Comment:
Why is this unreachable?

Reply:
get_interned_key is only supposed to be used when the key is known to already exist. But I'll make this a throw instead.